iot/mqtt_example/cloudiot_mqtt_example_nodejs.js

```diff
@@ -222,88 +352,470 @@ function publishAsync(messagesSent, numMessages) {
+        messagesSent + 1,
+        numMessages,
+        registryId,
+        deviceId,
+        gatewayId
+      );
+    }, schedulePublishDelayMs);
+  }, publishDelayMs);
+}
+
+// Sends data from a gateway on behalf of a device that is bound to that gateway.
+function sendDataFromBoundDevice(
+  gatewayId,
+  deviceId,
+  registryId,
+  projectId,
+  region,
+  algorithm,
+  privateKeyFile,
+  mqttBridgeHostname,
+  mqttBridgePort,
+  numMessages,
+  tokenExpMins
+) {
+  // [START iot_send_data_from_bound_device]
+  // const parentName = `projects/${projectId}/locations/${region}`;
+  // const registryName = `${parentName}/registries/${registryId}`;
+  // const binaryData = Buffer.from(data).toString('base64');
+  // const request = {
+  //   name: `${registryName}/devices/${deviceId}`,
+  //   binaryData: binaryData
+  // };
+
+  const mqttClientId = `projects/${projectId}/locations/${region}/registries/${registryId}/devices/${gatewayId}`;
+  console.log(`MQTT client id: ${mqttClientId}`);
+  let connectionArgs = {
+    host: mqttBridgeHostname,
+    port: mqttBridgePort,
+    clientId: mqttClientId,
+    username: 'unused',
+    password: createJwt(projectId, privateKeyFile, algorithm),
+    protocol: 'mqtts',
+    qos: 1,
+    secureProtocol: 'TLSv1_2_method',
+  };
+
+  // Create a client, and connect to the Google MQTT bridge.
+  let iatTime = parseInt(Date.now() / 1000);
+  let client = mqtt.connect(connectionArgs);
+
+  client.on('connect', success => {
+    if (!success) {
+      console.log('Client not connected...');
+    } else if (!publishChainInProgress) {
+      console.log('Client connected: Attaching device');
+      attachDevice(deviceId, client);
+      setTimeout(() => {
+        console.log('Client connected: Gateway is ready to relay');
+        publishAsyncGateways(
+          client,
+          iatTime,
+          tokenExpMins,
+          0,
+          numMessages,
+          registryId,
+          deviceId,
+          gatewayId,
+          connectionArgs,
+          projectId,
+          privateKeyFile,
+          algorithm
+        );
+      }, 5000);
+    }
+  });
+
+  client.on('close', () => {
+    console.log('Connection closed');
+    shouldBackoff = true;
+  });
+
+  client.on('error', err => {
+    console.log('error', err);
+  });
+
+  client.on('message', (topic, message) => {
+    console.log(
+      'message received: ',
+      Buffer.from(message, 'base64').toString('ascii')
+    );
+  });
+
+  client.on('packetsend', () => {
+    // Note: logging packet send is very verbose
+  });
+  // [END iot_send_data_from_bound_device]
+}
+
+// Listen for configuration messages on a gateway and bound device.
+function listenForConfigMessages(
+  gatewayId,
+  deviceId,
+  registryId,
+  projectId,
+  region,
+  algorithm,
+  privateKeyFile,
+  mqttBridgeHostname,
+  mqttBridgePort,
+  clientDuration
+) {
+  // [START iot_listen_for_config_messages]
+  // const parentName = `projects/${projectId}/locations/${region}`;
```

Comment:
As above, you can update this section with the parameters used in the function.

Reply:
Fixed.